Updates to terrain world component docs. (#1884)
o3de/o3de.org: content/docs/user-guide/components/reference/terrain/world.md

You must add this component to the *Level* entity, the parent of all entities in an **Open 3D Engine (O3DE)** level.
{{< /important >}}

## Usage 

The **Min Height** and **Max Height** bounds should be set to as small a range as possible for the most accurate height values. For instance, the terrain renderer uses 16 bit height values, which provides 65536 possible distinct heights. If the difference between the min and max height is 1km, that's a resolution of about 1.5cm.

**Height Query Resolution** is a synchronization point for terrain height used by different terrain systems like physics and rendering. For physics, this controls the density of the underlying height field collider. For rendering, this determines the triangle density of the closest level of detail.

> **Surface Data Query Resolution (m)** is similar to above, but it controls the resolution of surface data instead of heights. For instance, this affects the density of detail materials for detail material blending.

Reviewer: Suggestion: It might be worth describing a bit more how this conceptually creates two terrain world grids - one for height, and one for surface data. All height and surface data providers will be queried at this grid resolution, regardless of the resolution of the data in those providers, so the ideal goal is to use data that matches this resolution. For example, if a heightmap with 10 texels/m is used with a terrain that has a 1 m height query resolution, 9/10 of the texels will be ignored. If the heightmap has 1 texel/m with a 1 m height query resolution, then all of the texels will be used.

You must add this component to the *Level* entity, the parent of all entities in an **Open 3D Engine (O3DE)** level.
{{< /important >}}

## Usage 

> The **Min Height** and **Max Height** bounds should be set to as small a range as possible for the most accurate height values. For instance, the terrain renderer uses 16 bit height values, which provides 65536 possible distinct heights. If the difference between the min and max height is 1km, that's a resolution of about 1.5cm.

Reviewer 2: 16 bit -> 16 bits

Author: In this case "16 bit" is correct

Reviewer: Nit: Actually it should be "16-bit". "1km" and "1.5cm" should also have a space between the number and the unit. ("1 km", "1.5 cm")

**Height Query Resolution** and **Surface Data Query Resolution** are synchronization points for terrain height and surface data used by different terrain systems like physics and rendering. They effectively create grids across the world - one for height, and one for surface data. All height and surface data providers will be queried at this grid resolution regardless of the resolution of the data in providers. Ideally, source data in providers with quantized data (like image gradients) will match the query resolution used.

For example, if a heightmap with 10 texels/m is used with a terrain that has a 1m height query resolution, 9/10 of the texels will be ignored. If the heightmap has 1 texel/m with a 1 m height query resolution, then all of the texels will be used.

For physics, **Height Query Resolution** controls the density of the underlying height field collider. For rendering, this determines the triangle density of the closest level of detail.
**Surface Data Query Resolution** affects the density of detail materials for detail material blending.

> 

Reviewer: Suggestion: It would be nice to describe the usage of the CLOD settings here too. Specifically, that the "CLOD distance" should ideally be < 25% of "First LOD distance" and definitely should not be > 50% of "First LOD distance", or else the blend distance will be larger than the distance that the LOD renders for and the blend will never successfully complete, causing visual seams at the LOD boundaries.

The **Min Height** and **Max Height** bounds should be set to as small a range as possible for the most accurate height values. For instance, the terrain renderer uses 16 bit height values, which provides 65536 possible distinct heights. If the difference between the min and max height is 1km, that's a resolution of about 1.5cm.

> **Height Query Resolution** and **Surface Data Query Resolution** are a synchronization points for terrain height used by different terrain systems like physics and rendering. The effectively create grids across the world - one for height, and one for surface data. All height and surface data providers will be queried at this grid resolution regardless of the resolution of the data in providers. Ideally, source data in providers with quantized data (like image gradients) will match the query resolution used.

Reviewer: "are a" -> "are"
"points for terrain height" -> "points for terrain height and surface data"
"The effectively" -> "They effectively"

| Property | Description | Values | Default |
| - | - | - | - |
| Min Height | The minimum value for the height of terrain. The **min** value must be less than the **max** value.| -65536.0 to 65536.0 | 0.0 |
| Max Height | The maximum value for the worlds bounds. The **max** value must be greater than the **min** value. | -65536.0 to 65536.0 | 1024.0 |
| Height Query Resolution (m) | The distance between each height sample position, in meters. | 0.1 to Infinity | 1.0 |
| Surface Data Query Resolution (m) | The distance between each surface data sample position, in meters. | 0.1 to Infinity | 1.0 |

## Usage 

> The **Min Height** and **Max Height** bounds should be set to as small a range as possible for the most accurate height values. For instance, the terrain renderer uses 16-bit height values, which provides 65536 possible distinct heights. If the difference between the min and max height is 1 km, that's a resolution of about 1.5 cm.

Author: Your wording improves things :)

To try to explain it more, imagine a terrain with a min height of 0 and max height of 1000. With 65536 total possible distinct values to work with, that means with a 1000-meter range there are enough positions to have one distinct position every 1.5 cm. (The actual math is just 1000m / 65535 = 0.015259m). So, we could represent 0 perfectly, but 1cm would get rounded up to 1.5cm and 0.5cm would round down to 0cm. If your total height range was higher, like 10,000m (say you wanted to handle sea level to mount Everest) then you'd only have one distinct value per 15cm, so there'd be more rounding, and you might lose some fine detail.

> The **Min Height** and **Max Height** bounds should be set to as small a range as possible for the most accurate height values. For instance, the terrain renderer uses 16-bit height values, which provides 65536 possible distinct heights. If the difference between the min and max height is 1 km, that's a resolution of about 1.5 cm.

Reviewer 3: Cool! I incorporated some of your explanation into my suggestion: 
```suggestion
We recommend that you set the **Min Height** and **Max Height** bounds to as small of a range as possible,  to achieve more accurate height values . The terrain renderer uses 16-bit height values, which provides 65536 possible distinct heights. To get the resolution, calculate: (Max Height - Min Height) / 65536. So, if the height range is 1 km, then the resolution is about 1.5 cm. This results in finer height detail. 
```

> The **Min Height** and **Max Height** bounds should be set to as small a range as possible for the most accurate height values. For instance, the terrain renderer uses 16-bit height values, which provides 65536 possible distinct heights. If the difference between the min and max height is 1 km, that's a resolution of about 1.5 cm.

Reviewer 3: Cool! I incorporated this into my [latest suggestion](https://github.com/o3de/o3de.org/pull/1884#discussion_r942946590).

> We recommend that you set the **Min Height** and **Max Height** bounds to as small of a range as possible,  to achieve more accurate height values . The terrain renderer uses 16-bit height values, which provides 65536 possible distinct heights. So, if the difference between the min and max height is 1 km, then the resolution is about 1.5 cm. 

Reviewer 3: I incorporated your explanations into this suggestion: 

```suggestion
We recommend that you set the **Min Height** and **Max Height** bounds to as small of a range as possible,  to achieve more accurate height values . The terrain renderer uses 16-bit height values, which provides 65536 possible distinct heights. To get the resolution, calculate: (Max Height - Min Height) / 65536. So, if the height range is 1 km, then the resolution is about 1.5 cm. This results in finer height detail. 
```

> **Height Query Resolution** and **Surface Data Query Resolution** are synchronization points for terrain height and surface data used by different terrain systems like physics and rendering. They effectively create grids across the world - one for height, and one for surface data. All height and surface data providers will be queried at this grid resolution regardless of the resolution of the data in providers. Ideally, source data in providers with quantized data (like image gradients) will match the query resolution used.

Reviewer 3: What is meant by "synchronization points" and "providers"? "Providers" may be ambiguous. 

Similarly, "source data" may be ambiguous, so I updated the last sentence. Can you verify if that's right?

```suggestion
 **Height Query Resolution** and **Surface Data Query Resolution** are synchronization points for both terrain height and surface data. Different terrain systems, like physics and rendering, use them to create height and surface data grids across the world. All height and surface data providers are queried at this grid resolution, regardless of the resolution of the data in providers.  
 
We recommend that heightmaps or surface data that have quantized source data, like image gradients, match the Height Query Resolution.
```

Author: "Synchronization points" - The query resolutions tell other systems that rely on terrain how to query data for their system. I referred to them as synchronization points because it helps keep different terrain systems in sync. As an example, if the query resolution is 1m, then both rendering and physics will query the same points 1 meter apart for their data, and everything will like up - a ball will bounce off the terrain in the location that it looks like it should bounce. Without this property, we'd have to leave it up to the user to make sure they made everything sync up by themselves. So, they don't create actual grids across the world, it's just a way of understanding how they work (ie query in regular intervals in x and y according to the query resolution)

"Providers" - this just means any component which can respond to queries about heights or surface data - it "provides" data for the terrain system.

"Source data" I think your clarification works. Source data could provide to anything providing the height or surface information like an image or noise function.

Author: How about this:

> **Height Query Resolution** and **Surface Data Query Resolution** are synchronization points for both terrain height and surface data. All of the entities providing input height and surface data for the terrain system will be queried at this grid resolution regardless of the resolution of the data contained in those entities. We recommend that heightmaps or surface data that have quantized source data, like image gradients, match the Height Query Resolution.

Author: This actually provoked a long discussion with @mbalfour-amzn and this is what we landed on:

> **Height Query Resolution** and **Surface Data Query Resolution** are used to define consistent spacing, centered at the origin, for terrain height and surface queries across the entire world. Different systems that use terrain data, such as physics and rendering, can also use these conceptual grids to make implementation decisions about how to best use the terrain data. The input data to the terrain system can be authored or generated at any resolution, but the terrain system will by default query the input data only at points aligned to the grid. This creates a consistent view of the input data. Since the input data will normally only be used at these resolutions, we recommend that input gradients that have quantized source data, like images, match the appropriate query resolution.

Reviewer 3: Great! This makes a lot more sense. And "conceptual grid" is a good way to describe it.

```suggestion
**Height Query Resolution** and **Surface Data Query Resolution** are used to define consistent spacing, centered at the origin, for terrain height and surface queries across the entire world. Different systems that use terrain data, such as physics and rendering, can also use these conceptual grids to make implementation decisions about how to best use the terrain data. By default, the terrain system queries the input data only at points aligned to the grid. However, you can author or generate input data to the terrain system at any resolution. This creates a consistent view of the input data. Since the input data is only used at these resolutions, we recommend that input gradients that have quantized source data, like images, match the appropriate query resolution.
```

> 
> For example, if a heightmap with 10 texels/m is used with a terrain that has a 1m height query resolution, 9/10 of the texels will be ignored. If the heightmap has 1 texel/m with a 1 m height query resolution, then all of the texels will be used.

Reviewer 3: ```suggestion
For example, if a terrain uses a heightmap with 10 texels/meter with a 1 meter height query resolution, then 9 out of 10 of the texels are ignored. If the heightmap has 1 texel/meter with a 1 meter height query resolution, then all of the texels are used.
```

Author: After reading this again I'm going to change "heightmap" to "heightmap image" to be clearer since we're talking about an image source that happens to contain heightmap information.

| Property | Description | Values | Default |
| - | - | - | - |
| Min Height | The minimum value for the height of terrain. The **min** value must be less than the **max** value.| -65536.0 to 65536.0 | 0.0 |
| Max Height | The maximum value for the worlds bounds. The **max** value must be greater than the **min** value. | -65536.0 to 65536.0 | 1024.0 |
| Height Query Resolution (m) | The distance between each height sample position, in meters. | 0.1 to Infinity | 1.0 |
| Surface Data Query Resolution (m) | The distance between each surface data sample position, in meters. | 0.1 to Infinity | 1.0 |

{{< note >}}

> Because large numbers of terrain height samples can take a long time to calculate, O3DE will not allow the number of samples to exceed 8 million.

Reviewer: Is this note still accurate?

Author: No, I think it'll just crash now if you make the values too crazy. That's what we have an open UX ticket about. I'll remove this.
